src/frontend/src/optimizer/property/monotonicity.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use enum_as_inner::EnumAsInner;
+use risingwave_common::types::DataType;
+use risingwave_pb::expr::expr_node::Type as ExprType;
+
+use crate::expr::{Expr, ExprImpl, FunctionCall, TableFunction};
+
+/// Represents the monotonicity of a column. This enum aims to unify the "non-decreasing analysis" and watermark derivation.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EnumAsInner)]
+pub enum Monotonicity {
+    /// The column is constant.
+    Constant,
+    /// The column is strictly non-decreasing. Basically this is similar to `WATERMARK FOR col AS col`, but the `Watermark`
+    /// messages are not supposed to be produced by `WatermarkFilter`.
+    Increasing,
+    /// The column is strictly non-increasing.
+    Decreasing,
+    /// The column is GENERALLY non-decreasing, but not STRICTLY. Watermarks should be produced by `WatermarkFilter` and
+    /// forwarded to downstream with necessary transformation in other operators like `Project(Set)`.
+    /// This is not used currently.
+    _IncreasingByWatermark,
```

Comment:
```suggestion
    _IncreasingByWatermark(usize),
```
I have considered this situation but cannot provide an example. LGTM to keep it